File: src/ai/AssetLoader.ts
import fs from 'fs';
import path from 'path';

class AssetLoader {
  private static instance: AssetLoader | null = null;
  public assets: Record<string, string> = {};

  private constructor() {}

  static async initialize(basePath = process.cwd()) {
    if (AssetLoader.instance) return AssetLoader.instance;
    const loader = new AssetLoader();
    const modelDir = path.resolve(basePath, 'models');

    const files = {
      roles: 'roles.pkl',
      careerProfiles: 'career_profiles.pkl',
      roleSkillFreq: 'role_skill_freq.pkl',
      embeddings: 'arivium_role_embeddings.npy',
      faissIndex: 'arivium_faiss.index',
    };

    for (const [key, fname] of Object.entries(files)) {
      const p = path.join(modelDir, fname);
      if (!fs.existsSync(p)) {
        console.warn(`[AssetLoader] Warning: asset missing: ${p}`);
      }
      loader.assets[key] = p;
    }

    AssetLoader.instance = loader;
    return loader;
  }

  static getInstance(): AssetLoader {
    if (!AssetLoader.instance) throw new Error('AssetLoader not initialized');
    return AssetLoader.instance;
  }
}

export default AssetLoader;

File: backend/python/asset_loader.py
import sys
import json
from pathlib import Path
import pickle

# Usage: python asset_loader.py <project_root>
root = Path(sys.argv[1]) if len(sys.argv) > 1 else Path('.').resolve()
models_dir = root / 'models'

required = {
    # repo uses these filenames
    'roles': models_dir / 'arivium_roles.pkl',
    'career_profiles': models_dir / 'career_profiles.pkl',
    'role_skill_freq': models_dir / 'role_skill_freq.pkl',
    'embeddings': models_dir / 'arivium_role_embeddings.npy',
    'faiss': models_dir / 'arivium_faiss.index',
}

out = {
    'rolesPath': str(required['roles']),
    'careerProfilesPath': str(required['career_profiles']),
    'roleSkillFreqPath': str(required['role_skill_freq']),
    'embeddingsPath': str(required['embeddings']),
    'faissIndexPath': str(required['faiss']),
}

# Load pickles lightly to ensure they exist and are parseable
for k in ['roles', 'career_profiles', 'role_skill_freq']:
    p = required[k]
    if not p.exists():
        print(json.dumps({'error': f'missing {k} at {p}'}))
        sys.exit(2)
    try:
        with p.open('rb') as f:
            pickle.load(f)
    except Exception as e:
        print(json.dumps({'error': f'failed to load {p}: {e}'}))
        sys.exit(3)

# Ensure embeddings and faiss index exist
for k in ['embeddings', 'faissIndex']:
    if not Path(out[k + 'Path']).exists():
        print(json.dumps({'error': f'missing {k} at {out[k + "Path"]}'}))
        sys.exit(4)

print(json.dumps(out))
